### ISSUE TITLE
Safety of evaluated code and terminating code from python

### ISSUE THREAD

**Issue by samul-1** [PiotrDabkowski/Js2Py#231]
Since I will be using js2py to run code that is user-submitted, I'd like some input as to what the safest way to do this in js2py would be. I haven't been able to find any docs that address this.

Also, is there a way to call `eval_js` with a timeout parameter? If I `eval_js` an infinite loop, then my python code will stop working as well. I could allow code to run for no longer than a second (it probably doesn't need more anyway), but I don't know how to do this in js2py.

Any input is highly appreciated.

**Comment by satbekmyrza** [comment]
I was also looking for a `timeout` parameter in `eval_js` for somewhat similar scenario. Will be great to see it implemented.

**Comment by stautonico** [comment]
3 years later and this is still an interest. Even with threading, there is no real way to force the js interpreter to stop. User-entered code could run forever. I don't want to have to reset the entire program if a user enters an infinite loop.

**Comment by 4144** [comment]
probably only solution is to use some subprocess/pollexecutor to spawn process and kill it after timeout

**Comment by stautonico** [comment]
yep @4144. Multithreading was easier, but multiprocessing + RPC is probably the solution (for me at least)

**Comment by worstperson** [comment]
For me, it looks like killing the thread with a trace should work. It's also possible to inject a variable check in the body of translated iteration statements to conditionally call exit(). These are the functions ForStatement, ForInStatement, WhileStatement, and DoWhileStatement in js2py/translators/translating_nodes.py

An example with the while statement:
```
def WhileStatement(type, test, body):
    test_code = trans(test)
    loop_controller.enter()
    body_code = trans(body)
    loop_controller.leave()
    #result = 'while %s:\n' % test_code + indent(body_code)
    result = 'while %s:\n' % test_code + indent("if var.get('__KILL_THREAD'): exit()\n") + indent(body_code)
    return result
```

And the test code:
```
import threading
import js2py

context = js2py.EvalJs()
context.__KILL_THREAD = False

def runJS():
    context.execute("var a=0; while(true){ a = a + 1 }")

a = threading.Thread(target=runJS)
a.start()
a.join(1)
context.__KILL_THREAD = True
a.join()
print(context.a)
```

Even if it never releases the lock, we could always inject a time.sleep there the same way.
